WooCommerce/Classes/Extensions/Date+StartAndEnd.swift

```diff
@@ -0,0 +1,77 @@
+import Foundation
+
+extension Date {
+    // MARK: Day
+
+    /// Returns self's start of day in the given time zone.
+    func startOfDay(timezone: TimeZone) -> Date {
+        let calendar = createCalendar(timezone: timezone)
+        return calendar.startOfDay(for: self)
+    }
+
+    /// Returns self's end of day in the given time zone.
+    func endOfDay(timezone: TimeZone) -> Date {
+        let calendar = createCalendar(timezone: timezone)
+        var components = DateComponents()
+        components.day = 1
+        components.second = -1
+        return calendar.date(byAdding: components, to: startOfDay(timezone: timezone))!
+    }
+
+    // MARK: Week
+
+    /// Returns self's start of week in the given time zone.
+    func startOfWeek(timezone: TimeZone) -> Date {
+        let calendar = createCalendar(timezone: timezone)
+        let components = calendar.dateComponents([.yearForWeekOfYear, .weekOfYear], from: startOfDay(timezone: timezone))
+        return calendar.date(from: components)!
+    }
+
+    /// Returns self's end of week in the given time zone.
+    func endOfWeek(timezone: TimeZone) -> Date {
+        let calendar = createCalendar(timezone: timezone)
+        return calendar.date(byAdding: .second, value: 604799, to: startOfWeek(timezone: timezone))!
```

Comment:
I admit it. I did the math. I see you calculated the date right down to the last possible second 😅